enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,696 @@
+* CRD Compatibility Checker ensures that the version Hypershift upgrades to does not break CCAPIO, and therefore the management cluster.
+
+Hypershift asserts ownership of the `Machine` CRD by adding `machines.cluster.x-k8s.io` to CCAPIO's `UnmanagedAPIs`.
+This is the only operational change required by Hypershift.
+
+CCAPIO observes the change to UnmanagedAPIs, and ensures that a CompatibilityRequirement exists for every entry.
+In the case of `Machine` this would be:
+
+```yaml
+apiVersion:	apiextensions.openshift.io/v1alpha1
+kind:	CompatibilityRequirement
+metadata:
+  name:	cluster-api-machine-ccapio
+spec:
+  compatibilitySchema:
+    customResourceDefinition:
+      data: |
+        ...
+        <complete YAML document of Machine CRD from transport config map>
+        ...
+      type: YAML
+    excludedFields:
+    - path: "status.deprecated"
+    requiredVersions:
+      # The v1beta1 version was added because we observed an annotation on a Deployment in OpenStack's
+      # transport configmap indicating that it is using CAPI Machine v1beta1.
+      additionalVersions:
+      - v1beta1
+      defaultSelection: StorageOnly # Valid options are StorageOnly and AllServed.
+  customResourceDefinitionSchemaValidation:
+    action: Deny
+  objectSchemaValidation:
+    action: Deny
+    namespaceSelector:
+      matchLabels:
+        "kubernetes.io/metadata.name": "openshift-cluster-api"
+```
+
+CCAPIO marks itself Degraded if CompatibilityRequirement it owns becomes not Progressing, and:
+* not Admitted - transport config map is invalid
+* not Compatible - current CRD is incompatible with CCAPIO requirements
+
+Hypershift proceeds normally and upgrades the `Machine` CRD.
+Assuming that the CRD from version Y is compatible with the one for version X required by CCAPIO, the CRD is admitted and everything works as normal.
+
+#### CCAPIO prevents cluster upgrade if it would conflict with an existing CRD
+
+This example covers the scenario when CCAPIO is first enabled in OCP, but more generally if CCAPIO starts using a CRD which Hypershift is currently using.
+We want to prevent upgrade if a CRD currently in use would be incompatible with the version required by CCAPIO after upgrade.
+
+OCP 4.Y is the current version.
+OCP 4.Y+1 is the upgraded version.
+
+Some time during the OCP 4.Y lifecycle we add an additional set of transport config maps to 4.Y.z which contain the CRDs required by 4.Y+1.
+CCAPIO scans these, but does not load them.
+For each CRD it creates a new CompatibilityRequirement.
+e.g. for the `Machine` CRD it would create:
+
+```yaml
+apiVersion:	apiextensions.openshift.io/v1alpha1
+kind:	CompatibilityRequirement
+metadata:
+  name:	cluster-api-machine-ccapio
+spec:
+  compatibilitySchema:
+    customResourceDefinition:
+      data: |
+        ...
+        <complete YAML document of Machine CRD from transport config map>
+        ...
+      type: YAML
+    excludedFields:
+    - path: "status.deprecated"
```

Comment:
On this and the example above, we don't specify the versions on the excluded fields. Should we addd an example with versions specified?